source/extensions/filters/network/thrift_proxy/conn_manager.cc

```diff
@@ -243,45 +246,137 @@ FilterStatus ConnectionManager::ResponseDecoder::transportEnd() {
   return FilterStatus::Continue;
 }
 
+void ConnectionManager::ActiveRpcDecoderFilter::continueDecoding() {
+  const FilterStatus status = parent_.applyDecoderFilters(this);
+  if (status == FilterStatus::Continue) {
+    // All filters have been executed for the current decoder state.
+    if (parent_.pending_transport_end_) {
+      // If the filter stack was paused during transportEnd, handle end-of-request details.
+      parent_.finalizeRequest();
+    }
+
+    parent_.continueDecoding();
+  }
+}
+
+FilterStatus ConnectionManager::ActiveRpc::applyDecoderFilters(ActiveRpcDecoderFilter* filter) {
+  ASSERT(filter_action_ != nullptr);
+
+  if (!local_response_sent_) {
+    if (upgrade_handler_) {
+      // Divert events to the current protocol upgrade handler.
+      const FilterStatus status = filter_action_(upgrade_handler_.get());
+      filter_context_.reset();
+      return status;
+    }
+
+    std::list<ActiveRpcDecoderFilterPtr>::iterator entry;
+    if (!filter) {
+      entry = decoder_filters_.begin();
+    } else {
+      entry = std::next(filter->entry());
+    }
+
+    for (; entry != decoder_filters_.end(); entry++) {
+      const FilterStatus status = filter_action_((*entry)->handle_.get());
+      if (local_response_sent_) {
+        // The filter called sendLocalReply: stop processing filters and return
+        // FilterStatus::Continue irrespective of the current result.
+        break;
+      }
+
+      if (status != FilterStatus::Continue) {
+        return status;
+      }
+    }
+  }
+
+  filter_action_ = nullptr;
+  filter_context_.reset();
+
+  return FilterStatus::Continue;
+}
+
+FilterStatus ConnectionManager::ActiveRpc::transportBegin(MessageMetadataSharedPtr metadata) {
+  filter_context_ = metadata;
+  filter_action_ = [this](DecoderEventHandler* filter) -> FilterStatus {
+    MessageMetadataSharedPtr metadata = absl::any_cast<MessageMetadataSharedPtr>(filter_context_);
+    return filter->transportBegin(metadata);
+  };
+
+  return applyDecoderFilters(nullptr);
+}
+
 FilterStatus ConnectionManager::ActiveRpc::transportEnd() {
   ASSERT(metadata_ != nullptr);
-  ASSERT(metadata_->hasMessageType());
+
+  FilterStatus status;
+  if (upgrade_handler_) {
+    status = upgrade_handler_->transportEnd();
+
+    if (metadata_->isProtocolUpgradeMessage()) {
+      ENVOY_CONN_LOG(error, "thrift: sending protocol upgrade response",
```

Comment:
@zuercher 

Hello, I have a question about this.

Why did you change the level of this log sentence (`thrift: sending protocol upgrade response`) from debug to Error

This seems to have returned a Twitter upgrade response normally, not an error.